File: sdks/node/src/redact.ts
/**
 * The privacy floor: deterministic, dependency-light redaction that runs
 * **in-process before any bytes leave the SDK**.
 *
 * This is the TypeScript port of the Rust SDK's `redact.rs` (itself a port of
 * the daemon's `SECRET_FLOOR` plus the email / absolute-path PII rules). It is
 * the irreducible baseline — even in "raw" remote mode the floor still scrubs
 * live credentials; "raw" means *full turns*, not *leaked keys*.
 *
 * Placeholder style is SQUARE brackets `[REDACTED:name]` (matching the Rust
 * SDK — NOT the daemon's angle brackets). Rules are ordered specific → generic
 * so a known provider key is labelled precisely before the generic
 * env-secret / blob catchers run. Every pattern uses the `g` flag.
 *
 * Unlike Rust's `regex` crate, JS regex supports look-around, so the one
 * boundary-sensitive pattern (the 40-char AWS-secret blob) uses a clean
 * lookbehind/lookahead pair (Node ≥20) rather than explicit boundary capture
 * groups.
 */

/** Result of a redaction pass. */
export interface Redacted {
  /** The cleaned text. */
  text: string;
  /** Count of secret-format matches replaced. */
  secrets: number;
  /** Count of PII matches replaced (emails, absolute paths). */
  pii: number;
}

interface Rule {
  re: RegExp;
  /** Replacement string (may use `$1`/`$2` capture references). */
  replacement: string;
}

/**
 * Ordered specific → generic. Specific provider keys run before the generic
 * env-secret / blob catchers so a known key is labelled precisely.
 *
 * Each `RegExp` carries the `g` flag so `replace` swaps every occurrence and
 * `match` counts every occurrence. Patterns are recreated lazily on first use
 * (a module-level const array is fine — `lastIndex` is only consulted by
 * stateful `.exec`/`.test`, and we use `.match`/`.replace`, which reset it).
 */
const FLOOR: Rule[] = [
  // 1. Anthropic API keys.
  { re: /sk-ant-[A-Za-z0-9_-]{20,}/g, replacement: "[REDACTED:anthropic_key]" },
  // 2. OpenAI keys (incl. project-scoped `sk-proj-`).
  { re: /sk-(?:proj-)?[A-Za-z0-9_-]{20,}/g, replacement: "[REDACTED:openai_key]" },
  // 3. Google API keys.
  { re: /AIza[0-9A-Za-z_-]{35}/g, replacement: "[REDACTED:google_api_key]" },
  // 4. AWS access key ids.
  { re: /\b(?:AKIA|ASIA)[0-9A-Z]{16}\b/g, replacement: "[REDACTED:aws_access_key]" },
  // 5-7. GitHub PAT / OAuth / app tokens.
  { re: /ghp_[A-Za-z0-9]{36,}/g, replacement: "[REDACTED:github_pat]" },
  { re: /gho_[A-Za-z0-9]{36,}/g, replacement: "[REDACTED:github_oauth]" },
  { re: /gh[sur]_[A-Za-z0-9]{36,}/g, replacement: "[REDACTED:github_app]" },
  // 8. Slack tokens.
  { re: /xox[aboprs]-[A-Za-z0-9-]{10,}/g, replacement: "[REDACTED:slack_token]" },
  // 9-10. Stripe live / test keys.
  { re: /(?:sk|pk|rk)_live_[A-Za-z0-9]{24,}/g, replacement: "[REDACTED:stripe_live_key]" },
  { re: /(?:sk|pk|rk)_test_[A-Za-z0-9]{24,}/g, replacement: "[REDACTED:stripe_test_key]" },
  // 11. Discord bot tokens.
  { re: /[MN][A-Za-z\d]{23}\.[\w-]{6}\.[\w-]{27}/g, replacement: "[REDACTED:discord_token]" },
  // 12. JWTs (three base64url segments).
  {
    re: /eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}/g,
    replacement: "[REDACTED:jwt]",
  },
  // 13. PEM private-key blocks (multi-line; non-greedy body).
  {
    re: /-----BEGIN (?:RSA |EC |DSA |OPENSSH )?PRIVATE KEY-----[\s\S]*?-----END (?:RSA |EC |DSA |OPENSSH )?PRIVATE KEY-----/g,
    replacement: "[REDACTED:private_key]",
  },
  // 14. modelstat device secrets.
  { re: /ds_live_[A-Za-z0-9_-]{32,}/g, replacement: "[REDACTED:modelstat_device_secret]" },
  // 15. Generic env-style KEY=VALUE where KEY names a secret. KEEP the var name.
  {
    re: /\b([A-Z][A-Z0-9_]*(?:TOKEN|KEY|SECRET|PASSWORD|PASSWD|API)[A-Z0-9_]*)\s*[:=]\s*['"]?([^\s'"]{12,})['"]?/g,
    replacement: "$1=[REDACTED:env_secret]",
  },
  // 16. Bearer tokens.
  { re: /Bearer\s+[A-Za-z0-9._~+/-]{20,}=*/g, replacement: "Bearer [REDACTED:bearer]" },
  // 17. DB connection strings — drop the password, keep scheme + a `<user>` stub.
  {
    re: /\b(postgres|mysql|mongodb|redis|amqp)(?:\+[a-z]+)?:\/\/[^:\s]+:([^@\s]+)@/gi,
    replacement: "$1://<user>:[REDACTED:db_password]@",
  },
  // 18. (most generic, LAST among secrets) A lone 40-char base64-ish blob, e.g.
  //     an AWS secret access key. Look-around keeps an embedded blob inside a
  //     longer token untouched.
  {
    re: /(?<![A-Za-z0-9/+=])[A-Za-z0-9/+=]{40}(?![A-Za-z0-9/+=])/g,
    replacement: "[REDACTED:aws_secret_key]",
  },
];

// --- PII rules (after the secret floor) --------------------------------------

const EMAIL = /[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}/g;

/**
 * Absolute home paths on macOS / Linux / Windows — they leak usernames and
 * machine layout.
 */
const ABS_PATH = /(?:\/Users\/|\/home\/)[^\s"'`)]+|[A-Za-z]:\\Users\\[^\s"'`)]+/g;

/** Count how many times `re` (a `g`-flagged pattern) matches `text`. */
function countMatches(text: string, re: RegExp): number {
  const m = text.match(re);
  return m ? m.length : 0;
}

/**
 * Redact `input` against the floor. Returns the cleaned text and per-class
 * counts (counts are taken *before* the replacement, matching the Rust
 * reference). When nothing matches, the input is returned unchanged.
 */
export function redact(input: string): Redacted {
  let text = input;
  let secrets = 0;
  let pii = 0;

  for (const rule of FLOOR) {
    const n = countMatches(text, rule.re);
    if (n > 0) {
      text = text.replace(rule.re, rule.replacement);
      secrets += n;
    }
  }

  const emails = countMatches(text, EMAIL);
  if (emails > 0) {
    text = text.replace(EMAIL, "[REDACTED:email]");
    pii += emails;
  }

  const paths = countMatches(text, ABS_PATH);
  if (paths > 0) {
    text = text.replace(ABS_PATH, "[REDACTED:path]");
    pii += paths;
  }

  return { text, secrets, pii };
}

File: sdks/rust/src/redact.rs
//! The privacy floor: deterministic, dependency-light redaction that runs
//! **in-process before any bytes leave the SDK**.
//!
//! This is a Rust port of the daemon's `SECRET_FLOOR`
//! (`packages/core/src/redact-floor.ts`) plus the email / absolute-path PII
//! rules. It is the irreducible baseline — even in "raw" remote mode the floor
//! still scrubs live credentials; "raw" means *full turns*, not *leaked keys*.
//!
//! Parity note: Rust's `regex` crate has no look-around, so the one boundary-
//! sensitive pattern (the 40-char AWS-secret blob) is expressed with explicit
//! boundary capture groups instead of `(?<!…)`/`(?!…)`. Every other pattern is
//! a faithful port. The unit tests assert each credential family is caught.

use regex::Regex;
use std::sync::LazyLock;

/// Result of a redaction pass.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Redacted {
    pub text: String,
    /// Count of secret-format matches replaced.
    pub secrets: u32,
    /// Count of PII matches replaced (emails, absolute paths).
    pub pii: u32,
}

struct Rule {
    re: Regex,
    replacement: &'static str,
}

/// Ordered specific → generic. Specific provider keys run before the generic
/// env-secret / blob catchers so a known key is labelled precisely.
static FLOOR: LazyLock<Vec<Rule>> = LazyLock::new(|| {
    let r = |p: &str, repl: &'static str| Rule {
        re: Regex::new(p).expect("floor pattern compiles"),
        replacement: repl,
    };
    vec![
        r(r"sk-ant-[A-Za-z0-9_-]{20,}", "[REDACTED:anthropic_key]"),
        r(r"sk-(?:proj-)?[A-Za-z0-9_-]{20,}", "[REDACTED:openai_key]"),
        r(r"AIza[0-9A-Za-z_-]{35}", "[REDACTED:google_api_key]"),
        r(
            r"\b(?:AKIA|ASIA)[0-9A-Z]{16}\b",
            "[REDACTED:aws_access_key]",
        ),
        r(r"ghp_[A-Za-z0-9]{36,}", "[REDACTED:github_pat]"),
        r(r"gho_[A-Za-z0-9]{36,}", "[REDACTED:github_oauth]"),
        r(r"gh[sur]_[A-Za-z0-9]{36,}", "[REDACTED:github_app]"),
        r(r"xox[aboprs]-[A-Za-z0-9-]{10,}", "[REDACTED:slack_token]"),
        r(
            r"(?:sk|pk|rk)_live_[A-Za-z0-9]{24,}",
            "[REDACTED:stripe_live_key]",
        ),
        r(
            r"(?:sk|pk|rk)_test_[A-Za-z0-9]{24,}",
            "[REDACTED:stripe_test_key]",
        ),
        r(
            r"[MN][A-Za-z\d]{23}\.[\w-]{6}\.[\w-]{27}",
            "[REDACTED:discord_token]",
        ),
        r(
            r"eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}",
            "[REDACTED:jwt]",
        ),
        r(
            r"-----BEGIN (?:RSA |EC |DSA |OPENSSH )?PRIVATE KEY-----[\s\S]*?-----END (?:RSA |EC |DSA |OPENSSH )?PRIVATE KEY-----",
            "[REDACTED:private_key]",
        ),
        r(
            r"ds_live_[A-Za-z0-9_-]{32,}",
            "[REDACTED:modelstat_device_secret]",
        ),
        // Generic env-style KEY=VALUE where KEY names a secret. Keeps the name.
        r(
            r#"\b([A-Z][A-Z0-9_]*(?:TOKEN|KEY|SECRET|PASSWORD|PASSWD|API)[A-Z0-9_]*)\s*[:=]\s*['"]?([^\s'"]{12,})['"]?"#,
            "${1}=[REDACTED:env_secret]",
        ),
        r(
            r"Bearer\s+[A-Za-z0-9._~+/-]{20,}=*",
            "Bearer [REDACTED:bearer]",
        ),
        r(
            r"(?i)\b(postgres|mysql|mongodb|redis|amqp)(?:\+[a-z]+)?://[^:\s]+:([^@\s]+)@",
            "${1}://<user>:[REDACTED:db_password]@",
        ),
    ]
});

/// The 40-char base64-ish blob (e.g. a lone AWS secret access key). Boundary
/// groups stand in for the TS look-around so an embedded blob inside a longer
/// token is left alone.
static AWS_SECRET_BLOB: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"(^|[^A-Za-z0-9/+=])([A-Za-z0-9/+=]{40})([^A-Za-z0-9/+=]|$)")
        .expect("aws blob pattern compiles")
});

static EMAIL: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}").expect("email pattern compiles")
});

/// Absolute home paths on macOS / Linux / Windows — they leak usernames and
/// machine layout.
static ABS_PATH: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"(?:/Users/|/home/)[^\s"'`)]+|[A-Za-z]:\\Users\\[^\s"'`)]+"#)
        .expect("path pattern compiles")
});

/// Redact `input` against the floor. Returns the cleaned text and per-class
/// counts. Allocation-free when nothing matches (returns the input unchanged).
#[must_use]
pub fn redact(input: &str) -> Redacted {
    let mut text = input.to_string();
    let mut secrets: u32 = 0;
    let mut pii: u32 = 0;

    for rule in FLOOR.iter() {
        let n = rule.re.find_iter(&text).count() as u32;
        if n > 0 {
            text = rule.re.replace_all(&text, rule.replacement).into_owned();
            secrets += n;
        }
    }

    // AWS blob (boundary-preserving replacement).
    let n = AWS_SECRET_BLOB.find_iter(&text).count() as u32;
    if n > 0 {
        text = AWS_SECRET_BLOB
            .replace_all(&text, |c: &regex::Captures| {
                format!("{}[REDACTED:aws_secret_key]{}", &c[1], &c[3])
            })
            .into_owned();
        secrets += n;
    }

    let n = EMAIL.find_iter(&text).count() as u32;
    if n > 0 {
        text = EMAIL.replace_all(&text, "[REDACTED:email]").into_owned();
        pii += n;
    }

    let n = ABS_PATH.find_iter(&text).count() as u32;
    if n > 0 {
        text = ABS_PATH.replace_all(&text, "[REDACTED:path]").into_owned();
        pii += n;
    }

    Redacted { text, secrets, pii }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn clean(s: &str) -> String {
        redact(s).text
    }

    #[test]
    fn scrubs_each_secret_family() {
        let cases = [
            "sk-ant-0123456789abcdefghijABCDEF",
            "sk-proj-0123456789abcdefghijABCDEF",
            "AIzaSyA1234567890123456789012345678901234",
            "AKIAIOSFODNN7EXAMPLE",
            "ghp_0123456789012345678901234567890123456789",
            "xoxb-1234567890-abcdefghijkl",
            "sk_live_0123456789012345678901234567",
            "ds_live_0123456789012345678901234567890123",
        ];
        for c in cases {
            let out = clean(c);
            assert!(
                out.contains("[REDACTED:"),
                "expected redaction for {c:?}, got {out:?}"
            );
            assert!(!out.contains(&c[..c.len().min(12)]) || out.contains("REDACTED"));
        }
    }

    #[test]
    fn keeps_env_var_name_but_drops_value() {
        let out = clean("MY_API_TOKEN=supersecretvalue123");
        assert!(out.contains("MY_API_TOKEN="), "got {out:?}");
        assert!(out.contains("[REDACTED:env_secret]"), "got {out:?}");
        assert!(!out.contains("supersecretvalue123"));
    }

    #[test]
    fn redacts_bearer_and_db_password() {
        let b = clean("Authorization: Bearer abcdefghijklmnopqrstuvwxyz0123");
        assert!(b.contains("Bearer [REDACTED:bearer]"), "got {b:?}");
        let d = clean("postgres://app:hunter2hunter2@db.internal:5432/prod");
        assert!(d.contains("[REDACTED:db_password]"), "got {d:?}");
        assert!(!d.contains("hunter2hunter2"));
    }

    #[test]
    fn redacts_email_and_paths_as_pii() {
        let r = redact("ping me at jane.doe@example.com from /Users/jane/secret/app.rs");
        assert!(r.text.contains("[REDACTED:email]"), "got {:?}", r.text);
        assert!(r.text.contains("[REDACTED:path]"), "got {:?}", r.text);
        assert_eq!(r.pii, 2);
    }

    #[test]
    fn leaves_clean_text_untouched_and_counts_zero() {
        let r = redact("refactor the auth module and add tests");
        assert_eq!(r.text, "refactor the auth module and add tests");
        assert_eq!(r.secrets, 0);
        assert_eq!(r.pii, 0);
    }

    #[test]
    fn aws_secret_blob_is_caught() {
        // The canonical 40-char AWS secret-key example, standing alone.
        let key = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY";
        assert_eq!(key.len(), 40);
        let out = clean(&format!("aws_secret = {key}"));
        assert!(out.contains("[REDACTED:aws_secret_key]"), "got {out:?}");
        assert!(!out.contains(key));
    }
}
